src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/LagPortOperationService.java

```diff
@@ -213,45 +219,94 @@ public String getSwitchIpAddress(SwitchId switchId) throws InvalidDataException,
+        log.info(
+                "Updating LAG logical port #{} on {} entry desired target ports set {}, current target ports set {}",
+                logicalPortNumber, switchId,
+                formatPortNumbersSet(targetPorts), formatPortNumbersSet(targetsBeforeUpdate));
+        replacePhysicalPorts(port, switchId, targetPorts);
+
+
+        return targetsBeforeUpdate;
+    }
+
     private LagLogicalPort deleteTransaction(SwitchId switchId, int logicalPortNumber) {
         LagLogicalPort port = ensureDeleteIsPossible(switchId, logicalPortNumber);
         log.info("Removing LAG logical port entry into DB: {}", port);
         lagLogicalPortRepository.remove(port);
         return port;
     }
 
+    private void ensureLagDataValid(Switch sw, Set<Integer> targetPorts) {
+        ensureSwitchIsLagCapable(sw);
+        ensureTargetPortsValid(sw, targetPorts);
+    }
+
+    private void ensureSwitchIsLagCapable(Switch sw) {
+        if (!isSwitchLagCapable(sw)) {
+            throw new InvalidDataException(format("Switch %s doesn't support LAG.", sw.getSwitchId()));
+        }
+    }
+
     private boolean isSwitchLagCapable(Switch sw) {
         return sw.getFeatures().contains(SwitchFeature.LAG);
     }
 
+    private void ensureTargetPortsValid(Switch sw, Set<Integer> targetPorts) {
+        Set<SwitchFeature> features = sw.getFeatures();
+        SwitchId switchId = sw.getSwitchId();
+        for (Integer portNumber : targetPorts) {
+            validatePhysicalPort(switchId, features, portNumber);
+        }
+    }
+
     private void ensureNoLagCollisions(SwitchId switchId, Set<Integer> targetPorts) {
-        Set<Integer> occupiedPorts = physicalPortRepository.findPortNumbersBySwitchId(switchId);
+        ensureNoLagCollisions(switchId, targetPorts, null);
+    }
+
+    private void ensureNoLagCollisions(SwitchId switchId, Set<Integer> targetPorts, Integer excludeLogicalPort) {
         // FIXME(surabujin): we are unreasonably supposing that all physical port objects in DB related to LAGs
-        SetView<Integer> intersection = Sets.intersection(occupiedPorts, new HashSet<>(targetPorts));
+        Collection<PhysicalPort> occupiedPorts = physicalPortRepository.findBySwitchId(switchId);
+        Set<Integer> deniedTargets = new HashSet<>();
+        for (PhysicalPort entry : occupiedPorts) {
+            LagLogicalPort partOf = entry.getLagLogicalPort();
+            if (excludeLogicalPort != null && partOf != null && excludeLogicalPort == partOf.getLogicalPortNumber()) {
+                log.debug(
+                        "Exclude port #{} on {} from collision list, because it is part of {} LAG logical port",
+                        entry.getPortNumber(), switchId, excludeLogicalPort);
+                continue;
+            }
+            deniedTargets.add(entry.getPortNumber());
+        }
+
+        SetView<Integer> intersection = Sets.intersection(deniedTargets, new HashSet<>(targetPorts));
```

Comment:
As for me something like this is more readable but it's a matter of taste :)

```
Optional<Integer> exclude = Optional.ofNullable(excludeLogicalPort);
Set<Integer> collision = occupiedPorts.stream()
                .filter(port -> port.getLagLogicalPort() != null)
                .filter(port -> !exclude.isPresent() || port.getLagLogicalPort().getLogicalPortNumber() != exclude.get())
                .map(PhysicalPort::getPortNumber)
                .filter(targetPorts::contains)
                .collect(Collectors.toSet());
```

Reply:
Can't agree. It brinngs too many excess logic into resulting code.